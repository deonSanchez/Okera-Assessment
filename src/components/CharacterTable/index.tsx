import {
  Table,
  TableHead,
  TableCell,
  TableBody,
  TableRow,
  Grid,
  Button,
  Typography,
} from '@mui/material';
import { useNavigate, createSearchParams } from 'react-router-dom';
import useStyles from './styles';

const MAX_PAGES = 214;
const MIN_PAGES = 1;

interface CharacterTableProps {
  page: string;
  data: any;
  loading?: boolean;
  hidden: boolean;
}

const CharacterTable = ({
  page,
  data,
  hidden,
  loading = false,
}: CharacterTableProps): JSX.Element => {
  const classes = useStyles();
  let pagination = parseInt(page) || 1;
  let navigate = useNavigate();

  const handleFirst = () => {
    navigate(`/character/${MIN_PAGES}`);
  };

  const handleLast = () => {
    navigate(`/character/${MAX_PAGES}`);
  };

  const handleNext = () => {
    pagination += 1;
    navigate(`/character/${pagination}`);
  };

  const handlePrevious = () => {
    pagination -= 1;
    navigate(`/character/${pagination}`);
  };

  const handleLink = (personUrl: string, pagination: number) => {
    const personId = personUrl.split('/').pop();

    navigate({
      pathname: `/character/${pagination}/details`,
      search: createSearchParams({ id: personId }).toString(),
    });
  };

  return (
    <Grid container direction="column">
      <Grid item>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell>Name (Alias) </TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {loading && (
              <TableRow>
                <TableCell>Loading Data...</TableCell>
              </TableRow>
            )}
            {data?.map((person) => (
              <TableRow key={person.url}>
                <TableCell
                  className={classes.tableCell}
                  onClick={() => handleLink(person.url, pagination)}
                >
                  {person.aliases[0] !== ''
                    ? `${person.name} (${person.aliases[0]})`
                    : `${person.name}`}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </Grid>
      <Grid item>
        {!hidden ? (
          <Grid
            container
            justifyContent="space-between"
            alignItems="center"
            spacing={3}
          >
            <Grid item>
              <Button
                className={classes.button}
                data-testid="first-page-button"
                disabled={pagination === MIN_PAGES}
                sx={{ marginRight: 2 }}
                onClick={handleFirst}
                variant="contained"
                size="large"
              >
                First Page
              </Button>
              <Button
                className={classes.button}
                data-testid="previous-page-button"
                disabled={pagination === MIN_PAGES}
                onClick={handlePrevious}
                variant="contained"
                size="large"
              >
                Previous Page
              </Button>
            </Grid>
            <Grid item>
              <Typography
                className={classes.typography}
                data-testid="pagination-text"
              >
                Page {pagination} of {MAX_PAGES}
              </Typography>
            </Grid>
            <Grid item>
              <Button
                className={classes.button}
                data-testid="next-page-button"
                disabled={pagination === MAX_PAGES}
                sx={{ marginRight: 2 }}
                onClick={handleNext}
                variant="contained"
                size="large"
              >
                Next Page
              </Button>
              <Button
                className={classes.button}
                data-testid="last-page-button"
                disabled={pagination === MAX_PAGES}
                onClick={handleLast}
                variant="contained"
                size="large"
              >
                Last Page
              </Button>
            </Grid>
          </Grid>
        ) : (
          <Grid container item justifyContent="flex-end">
            <Button
              className={classes.button}
              data-testid="back-character-button"
              onClick={handleFirst}
              variant="contained"
              size="large"
            >
              Back To Characters
            </Button>
          </Grid>
        )}
      </Grid>
    </Grid>
  );
};

export default CharacterTable;
