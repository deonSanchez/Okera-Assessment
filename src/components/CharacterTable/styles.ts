import { makeStyles } from '@mui/styles';

const useStyles = makeStyles(() => ({
  button: {
    marginTop: 16,
    textTransform: 'none',
  },
  tableCell: {
    '&:hover': {
      cursor: 'pointer',
      backgroundColor: '#f6f8fa',
    },
  },
  typography: {
    marginTop: 18,
  },
}));

export default useStyles;
