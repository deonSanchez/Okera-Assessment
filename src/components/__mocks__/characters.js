export const mockCharacter = {
  url: 'test-url',
  name: 'Deon Sanchez',
  gender: 'Male',
  culture: 'test-culture',
  born: 'test-born',
  died: 'test-died',
  titles: ['test-title', 'test-title2'],
  aliases: ['test-aliases', 'test-aliases22'],
  father: 'test-father',
  mother: 'test-mother',
  spouse: 'test-spouse',
  allegiances: ['test-allegiances', 'test-allegiances2'],
  books: ['test-books', 'test-books2'],
  povBooks: ['test-povBooks', 'test-povBooks2'],
  tvSeries: ['test-tvSeries', 'test-tvSeries2'],
  playedBy: ['test'],
};

export const mockCharacterTable = [mockCharacter];
