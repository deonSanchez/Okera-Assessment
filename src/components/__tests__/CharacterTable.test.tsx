import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import CharacterTable from 'components/CharacterTable';
import { mockCharacterTable } from '../__mocks__/characters';
import { MemoryRouter } from 'react-router-dom';

describe('CharacterTable', () => {
  test('should render CharacterTable', () => {
    const { container } = render(
      <CharacterTable page="3" data={mockCharacterTable} hidden={false} />,
      { wrapper: MemoryRouter }
    );

    screen.getByTestId('first-page-button');
    screen.getByTestId('last-page-button');
    screen.getByTestId('next-page-button');
    screen.getByTestId('previous-page-button');

    expect(container).toMatchSnapshot();
  });

  test('should hide pagination buttons', () => {
    render(
      <CharacterTable page="3" data={mockCharacterTable} hidden={true} />,
      { wrapper: MemoryRouter }
    );

    const button = screen.getByTestId('back-character-button');
    expect(button).toBeInTheDocument;
  });

  test('should check next page button functionality', () => {
    render(
      <CharacterTable page="3" data={mockCharacterTable} hidden={false} />,
      { wrapper: MemoryRouter }
    );

    const button = screen.getByTestId('next-page-button');
    const pagination = screen.getByTestId('pagination-text');

    expect(pagination).toHaveTextContent('Page 3 of 214');
    fireEvent.click(button);

    waitFor(() => {
      expect(pagination).toHaveTextContent('Page 4 of 214');
    });
  });
});
