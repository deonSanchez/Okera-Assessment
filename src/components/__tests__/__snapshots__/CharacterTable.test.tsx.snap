// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterTable should render CharacterTable 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Name (Alias) 
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium makeStyles-tableCell-2 css-1ex1afd-MuiTableCell-root"
            >
              Deon Sanchez (test-aliases)
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-n3lpn7-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root makeStyles-button-1 css-ju0db2-MuiButtonBase-root-MuiButton-root"
            data-testid="first-page-button"
            tabindex="0"
            type="button"
          >
            First Page
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root makeStyles-button-1 css-jh47zj-MuiButtonBase-root-MuiButton-root"
            data-testid="previous-page-button"
            tabindex="0"
            type="button"
          >
            Previous Page
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 makeStyles-typography-3 css-ahj2mt-MuiTypography-root"
            data-testid="pagination-text"
          >
            Page 
            3
             of 
            214
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root makeStyles-button-1 css-ju0db2-MuiButtonBase-root-MuiButton-root"
            data-testid="next-page-button"
            tabindex="0"
            type="button"
          >
            Next Page
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root makeStyles-button-1 css-jh47zj-MuiButtonBase-root-MuiButton-root"
            data-testid="last-page-button"
            tabindex="0"
            type="button"
          >
            Last Page
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
